### DISCUSSION TITLE
Technical Glossary

### DISCUSSION THREAD

**Discussion by flatcap** [neomutt/neomutt#3265] [Development]
I aim to define / distinguish:

- [x] Window, Dialog, Container, Panel, Bar, AllDialogsWindow, RootWindow
- [ ] Config set, subset, hash table, hash items, config definition
- [ ] Core object: NeoMutt, Account, Mailbox, Email (Envelope, Body)
- [ ] Sets: AccountList, MailboxList, EmailList, AccountArray, etc
- [ ] Notifications: sender, observer, event, type, notification, payload
- [ ] Gfx: index, menu, pager, sidebar, browser, helpbar, message window, index bar(status), pager bar
- [ ] Xdata: adata, mdata, edata, wdata, mdata (menu), inheritance
- [ ] Sets2: TAILQ, STAILQ, ARRAY, custom, node, ListHead, ListNode
- [ ] Data vs view, Context
- [ ] Function, opcode, command, dispatcher
- [ ] API, callback

**Comment by flatcap** [comment]
### Window, Dialog, Container, Panel, Bar, AllDialogsWindow, RootWindow

NeoMutt's screen is made up of a nested set of windows.
They are given lots of different names, but they're all of type `struct MuttWindow`.

<img align="right" width="480" src="https://raw.githubusercontent.com/neomutt/gfx/main/screenshots/window/dlg-index-pager-sidebar.svg">

**Window** - Any of the window types below
A Window may hold data.
A Window may be interactive.

**Dialog** - A set of windows that have a high-level function, e.g. Alias Selection Dialog.
(See `(D)` in the diagram)
A Dialog may hold data that's shared across all of its children.
A Dialog will have interactive parts.

**Container** - A non-visible window used for shaping groups of windows.
(See `(C)` in the diagram)
A Container may hold data.
A Container is not interactive.

**Panel** - This is a sub-division of a Dialog, it holds a group of windows that might be enabled/disabled as a set, e.g. Index Panel.
(See `(I)` in the diagram)
A Panel may hold data that's shared across all of its children.
A Panel may have interactive parts.

**Bar** - A Bar, or Status Bar, is a single line window
A Bar may hold data.
A Bar is not interactive.

**`AllDialogsWindow`** - A special named window.  This is the direct parent of all Dialogs.  Only one Dialog is ever visible.
(See `(A)` in the diagram)

**`RootWindow`** - A special named window.  This is the parent of the entire tree of windows.
(See `(R)` in the diagram)

**Reply by gahr** [reply]
Do we really need the distinction between Dialog and Panel?

**Reply by flatcap** [reply]
Need?  Certainly not, but...

Ignoring dialogs, there are only two **sets** of windows that we care about.
- Index Panel
- Pager Panel

Each can be shown/hidden as a set and they hold data that's shared within the set.

(It was getting tricky to distinguish the names of all the windows in the index/pager code :-)
